pkg/agent/multicluster/stretched_networkpolicy_controller.go

```diff
@@ -0,0 +1,426 @@
+			// the controller can only update the Pod classifier flow when the Pod container
+			// config is available. Events from the Informer may be received way before the Pod
+			// container config is available, which will cause the work item be continually
+			// re-queued with an exponential increased delay time. When the Pod container
+			// config is ready, the work item could wait for a long time to be processed.
+			UpdateFunc: controller.processPodUpdate,
+			DeleteFunc: controller.processPodDelete,
+		},
+		resyncPeriod,
+	)
+	controller.namespaceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: controller.processNamespaceUpdate,
+		},
+		resyncPeriod,
+	)
+	controller.labelIdentityInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.processLabelIdentityAddOrUpdate(cur)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.processLabelIdentityAddOrUpdate(cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.processLabelIdentityDelete(old)
+			},
+		},
+		resyncPeriod,
+	)
+	podUpdateSubscriber.Subscribe(controller.processPodCNIAddEvent)
+	return controller
+}
+
+func (s *StretchedNetworkPolicyController) Run(stopCh <-chan struct{}) {
+	defer s.queue.ShutDown()
+
+	klog.InfoS("Starting controller", "controller", stretchedNetworkPolicyControllerName)
+	defer klog.InfoS("Shutting down controller", "controller", stretchedNetworkPolicyControllerName)
+	cacheSyncs := []cache.InformerSynced{s.podListerSynced, s.namespaceListerSynced, s.LabelIdentityListerSynced}
+	if !cache.WaitForNamedCacheSync(stretchedNetworkPolicyControllerName, stopCh, cacheSyncs...) {
+		return
+	}
+	s.initLabelIDCache()
+	s.initPodClassifierFlow()
+	for i := 0; i < stretchedNetworkPolicyWorker; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (s *StretchedNetworkPolicyController) initLabelIDCache() {
+	labelIdentities, err := s.labelIdentityLister.List(labels.Everything())
+	if err == nil {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		for _, labelIdentity := range labelIdentities {
+			s.labelIdentityCache[labelIdentity.Spec.Label] = labelIdentity.Spec.ID
+		}
+	}
+}
+
+func (s *StretchedNetworkPolicyController) initPodClassifierFlow() {
+	pods, err := s.podLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Can't get Pods")
+		return
+	}
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		podNS, err := s.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			klog.ErrorS(err, "Can't get the Namespace of a Pod", "name", pod.Name, "namespace", pod.Namespace)
+			return
+		}
+		normalizedLabel := multicluster.GetNormalizedLabel(podNS.Labels, pod.Labels, podNS.Name)
+		classifierFlowUpdate := s.getClassifierFlowUpdate(getPodReference(pod), normalizedLabel)
+		s.queue.Add(classifierFlowUpdate)
```

Comment:
Using pod + labelID as the item of the queue makes it possible that a Pod is handled by more than one worker concurrently, one worker uses a newer ID while the other uses an older ID, but their execution order can not be guaranteed even if the newer one is enqueued earlier.

And this model is not very effecient and redundant as all event sources need to calculate labelID.

I wonder if we could just use Pod namespaced name as the item of queue, pros as below:
1. all event handlers just need to identify Pods that may be affected and enqueue them, no need to calculate labelIdentity on publisher side, only the sync func needs to do it.
2. race condition between multiple workers is avoided as a Pod will only be handled by a worker at any moment.

For Pod event handlers, they should just enqueue the Pod in the event.
For Namespace event handlers, they should just enqueue Pods in the namespace.
For LabelIdentity event handlers, they should just enqueue Pods using this label.

Reply:
I see your point. Although the enqueue order is correct, it's still possible that the stale LabelID overwrites the latest LabelID. But sorry I didn't understand why your suggestion could solve this issue. Hope you could share more details about pros 2. Thanks.

Reply:
Does this make workers meaningless? Because each worker needs to check the labelIdentity, which will block each other.

Reply:
The worker does more than checking labelidentity, right? Querying a map in memory is lightweight and making external call is heavy. I think we should serialize the former only?

Reply:
Make sense. Changed. Thanks!